lib/Builtins.cpp

```diff
@@ -14,171 +14,339 @@
+
+static bool GetUnmangledName(const std::string& str, size_t& pos, std::string& name) {
+  char* end;
+  int nameLen = strtol(&str[pos], &end, 10);
+  if (!nameLen) {
+    return false;
+  }
+  pos = end - str.data();
+
+  name = str.substr(pos, nameLen);
+  pos += nameLen;
+  return true;
 }
+
+static bool GetParameterType(clspv::Builtins::ParamTypeInfo& ti, const std::string& name, size_t& pos) {
+  // Parse a parameter type encoding
+  char typeCode = name[pos++];
+
+  int blen = 1;
+  switch (typeCode) {
+    // qualifiers
+  case 'P':                         // Pointer
+  case 'R':                         // Reference
+    return GetParameterType(ti, name, pos);
+  case 'k':                         // ??? not part of cxxabi
+  case 'K':                         // const
+  case 'V':                         // volatile
+    return GetParameterType(ti, name, pos);
+  case 'U': {                       // Address space
+    std::string address_space;
+    if (!GetUnmangledName(name, pos, address_space)) {
+      return false;
+    }
+    return GetParameterType(ti, name, pos);
+  }
+
+  case 'D':
+    typeCode = name[pos++];
+    if (typeCode == 'v') {        // OCL vector
+      char* end;
+      int numElems = strtol(&name[pos], &end, 10);
+      if (!numElems) {
+        return false;
+      }
+      ti.vsiz = numElems;
+      pos = end - name.data();
+
+      if (name[pos++] != '_') {
+        return false;
+      }
+      return GetParameterType(ti, name, pos);
+    } else if (typeCode == 'h') { // OCL half
+      ti.type      = Type::FloatTyID;
+      ti.sign      = true;
+      ti.blen      = 2;
+    } else {
+      assert(0);
+    }
+    break;
+
+    // element types
+  case 'l': blen <<= 1;             // long
+  case 'i': blen <<= 1;             // int
+  case 's': blen <<= 1;             // short
+  case 'c':                         // char
+  case 'a':                         // signed char
+    ti.type      = Type::IntegerTyID;
+    ti.sign      = true;
+    ti.blen      = blen;
+    break;
+  case 'm': blen <<= 1;             // unsigned long
+  case 'j': blen <<= 1;             // unsigned int
+  case 't': blen <<= 1;             // unsigned short
+  case 'h':                         // unsigned char
+    ti.type      = Type::IntegerTyID;
+    ti.sign      = false;
+    ti.blen      = blen;
+    break;
+  case 'd': blen   = 2;            // double float
+  case 'f': blen <<= 2;            // single float
+    ti.type      = Type::FloatTyID;
+    ti.sign      = true;
+    ti.blen      = blen;
+    break;
+  case 'v': // void
+    break;
+  case 'S':
+    if (name[pos++] != '_') {
+      return false;
+    }
+    break;
+  case '1':
+  case '2':
+  case '3':
+  case '4':
+  case '5':
+  case '6':
+  case '7':
+  case '8':
+  case '9': {
+    ti.type = Type::StructTyID;
+    pos--;
+    if (!GetUnmangledName(name, pos, ti.name)) {
+      return false;
+    }
+    break;
+  }
+  case '.':
+    return false;
+  default:
+#ifdef DEBUG
```

Comment:
I prefer an llvm_unreachable here rather than assert(0).
I want this to trigger in a release build as well.   Don't make it conditional on DEBUG

Reply:
Well it is permissible to fail to demangle if the name is not actually a mangled name and some names are an aggregate of strings, usually with a '.' delimiter (see above case) but is it always the case? IDK